packages/fan-fix-data-model/firebase/FirebaseAppUserModel.ts

```
@@ -28,21 +43,65 @@ export class FirebaseAppUserModel extends AppUserModel {
     );
   });
 
-  private readonly dataManager: FirebaseDataManager<AppUser>;
+  private readonly appUserManager: FirebaseDataManager<AppUser>;
+  private readonly usernameManager: FirebaseDataManager<Username>;
   
   constructor(
-    db: UnifiedFirestoreTypes.Module,
+    private readonly db: UnifiedFirestoreTypes.Module,
     private readonly staticClasses: UnifiedFirestoreTypes.StaticClasses,
     ) {
       super();
-      this.dataManager = new FirebaseDataManager(
+      this.appUserManager = new FirebaseDataManager(
         db.collection("AppUser"),
         deserializeAppUser,
-        { observeMode: "fetch" }
+        { observeMode: "fetch" },
       );
+      this.usernameManager = new FirebaseDataManager(
+        db.collection("Username"),
+        deserializeUsername,
+        { observeMode: "fetch" },
+      )
+  }
+
+  public serializeUpdateAppUserProfileParams = (params: UpdateAppUserProfileParams) => {
+    const { bio, email, realName, websiteLink } = params;
+    return {
+      email,
+      realName,
+      bio,
+      websiteLink,
+    }
+  };
+
+  public async isUsernameTaken(username: string): Promise<boolean> {
+    const result = await this.usernameManager.fetch(sanitizeUsername(username));
+    return !!result;
+  }
+
+  public async updateUsername(params: UpdateUsernameParams): Promise<void> {
+    const userId = params.id;
+    const username = sanitizeUsername(params.username);
+    
+    const usernameRef = this.usernameManager.collection.doc(sanitizeUsername(username));
+    const profileRef = this.appUserManager.collection.doc(userId);
+
+    await this.db.runTransaction(async (transaction) => {
+      const usernameSnapshot = await transaction.get(usernameRef);  
+
+      if (usernameSnapshot.exists && usernameSnapshot.data()?.uid !== userId) {
+        throw new Error(profileNameTaken);
```

hmm i feel a little weird throwing an error here. is it possible to just return undefined when unsuccessful, then if the username was successful? i just feel like throwing an error here requires someone calling the function to then wrap in a try/catch and handle the error and then they are limited with what they can do with it/may not even want to throw an error (maybe they just want to show a different message or something?). So my thinking is handling different return values is a bit more versatile. But maybe im off, thoughts?

---

I like that you're thinking about versatility and not having to wrap the function calls in try/catch. In this case, I do want to throw a this specific custom error because then we are being explicit about this function not completing due to this verbose condition. I think if we return undefined/value, it's not very obvious to me what undefined means, and having a return value itself may not be meaningful inside the context of an update function in general. My opinion is that it's a good practice to have all functions in the top of the call stack be in a try/catch block to handle errors anyway, and this custom error will allow them to cherry pick it to be handled differently, or be treated as a general error if no specific handling is required.

---

haha i see your point..hmm we seem to be at a stalemate! loll @mattryavec thoughts?